src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1381,6 +1386,96 @@
   (let [last-modified-milliseconds (.lastModified (io/file cacrl))]
        (time-coerce/from-long last-modified-milliseconds)))
 
+(schema/defn ^:always-validate reject-delta-crl
+  [crl :- CertificateRevocationList]
+  (when (utils/delta-crl? crl)
+    (throw (IllegalArgumentException.
+            ^String (i18n/trs "Cannot support delta CRL.")))))
+
+(schema/defn ^:always-validate validate-certs-and-crls
+  "Given a list of certificates and a list of CRLs, validate the certificate
+   chain, i.e. ensure that none of the certs have been revoked by checking the
+   appropriate CRL, which must be present and currently valid. Delta CRLs are
+   not supported. Returns nil if successful."
+  [cert-chain :- [Certificate]
+   crl-chain :- [CertificateRevocationList]]
+  (doseq [crl crl-chain]
+    (reject-delta-crl crl))
+  (try
+    (utils/validate-cert-chain cert-chain crl-chain)
+    ;; currently not distinguishing between invalid CRLs
+    ;; and a revoked cert in the CA chain
+    (catch CertPathValidatorException e
+      (throw (IllegalArgumentException.
+              ^String (i18n/trs "Invalid certs and/or CRLs: {0}" (.getMessage e)))))))
+
+(schema/defn ^:always-validate get-newest-crl :- CertificateRevocationList
+  "Determine the newest CRL by looking for the highest CRL Number. This assumes
+  all given CRLs have the same issuer. Fails if more than one CRL has the
+  highest CRL Number."
+  [crls :- [CertificateRevocationList]]
+  (let [newest-crl (->> crls
+                        (group-by utils/get-crl-number)
+                        (into (sorted-map))
+                        last
+                        last)]
+    (if (= 1 (count newest-crl))
+      (first newest-crl)
+      (throw (IllegalArgumentException.
+              ^String (i18n/trs "Could not determine newest CRL."))))))
+
+(schema/defn ^:always-validate maybe-replace-crl :- CertificateRevocationList
+  "Given a CRL and a map of key identifiers to CRLs, determine the
+  newest CRL with the key-id of the given CRL. Warn if the newest CRL
+  is the given CRL. Never replaces the CRL corresponding to the Puppet
+  CA signing cert."
+  [crl :- CertificateRevocationList
+   key-crl-map :- {KeyIdExtension [CertificateRevocationList]}
+   ca-cert-key-id :- [Byte]]
+  (let [key-id-ext (utils/get-extension-value crl utils/authority-key-identifier-oid)
+        maybe-new-crls (get key-crl-map key-id-ext)]
+    (if (= (:key-identifier key-id-ext) ca-cert-key-id)
```

Comment:
Would it be easier to use https://github.com/puppetlabs/jvm-ssl-utils/blob/main/src/clojure/puppetlabs/ssl_utils/core.clj#L616 ?

Reply:
Easier, yes, but I think that method is also wrong for the same reason that what I was doing here before was wrong, right? As in, it's going by name and not key ID. Are you suggesting we update that method to be better?